---
title: "coverage_workflow"
author: "Pol"
date: "2023-05-13"
output: html_document
---

## Spot low coverage positions
Low coverage was defined as **> 10%** of the species having no coverage for that particular position/site.
The coverage of the positions was obtained from the files in `file_PATH = /ANBpol/bc_Pathways/${puc}/pos_selection/positions_BT/${gen}_${position}.txt`. On this file, a '0' on the third column corresponds to those positions that both genomes and transcriptomes are not covered.

```{bash, eval=FALSE}
#!/bin/bash
path_in=/ANBpol/bc_Pathways/${puc}/pos_selection/positions_BT/
path_out=/ANBpol/Positive_selection_final/

for file in $path_in/*[0-9].txt ; do 
    out=$( cat $file | cut -f3 | grep '0' | grep -v 'position' | wc -l ) 
    echo -e "$file\t$out"
done > ${puc}_0-cov.txt
```
Only the genes with positions with low coverage were kept for further analysis.


## Plot coverage of low-coverage genes
```{r}
library("ggplot2")
spp_tr = 72
spp_gn = 508

gen = "ACTB_h1" # Gene to plot
sites=c("12482241", "12482245", "12482527", "12482783") # Sites under positive selection in the gene. Input manually

df_gn = read.table(paste("coverage_example_data/", gen,"_overall_cov.txt", sep = ""), sep = "\t", header = T)

hline_gn = (10*spp_gn)/100 # Line of 10% coverage
vline = sites # Plot vertical lines of sites

gn_plot <- ggplot(df_gn, aes(x=as.factor(pos), y=df_gn[,3])) +
  geom_bar(stat="identity") +
  geom_hline(yintercept = hline_gn, linetype="dashed", color = "red") +
  geom_vline(xintercept = vline, linetype="dashed", color = "blue") +
  geom_text(aes(800,hline_gn,label = "10%", vjust = -1, color = "red"), show.legend = F)
gn_plot
```

## Obtain intervals with low coverage
The following function does not return the interval if this reaches the end of the sequence

```{r}
# Function to extract intervals with low coverage
covint <- function(bins, threshold) {
  intervals <- list()
  i <- 1
  while (i <= length(bins)) {
    bin_i <- bins[i]
    if (bin_i > threshold) {
      j <- i
      while (j <= length(bins)) {
        bin_j <- bins[j]
        if (bin_j <= threshold) {
          intervals[[length(intervals) + 1]] <- c(i, j - 1)
          i <- j
          break
        }
        j <- j + 1
      }
    }
    i <- i + 1
  }
  return(intervals)
  
}

interv <- covint(df_gn[,2], 10)

#Transform index to intervals
df_out <- data.frame()
for (i in interv) {
  a = df_gn[i[1], 1]
  b = df_gn[i[2], 1]
  c = c(a,b)
  df_out <- rbind(df_out, t(data.frame(c)))
}

write.table(file = paste("coverage_example_data/", gen, "_intervals_discarded.txt", sep = ""), x = df_out, col.names = F, row.names = F, sep = "\t")
```

After obtaining the intervals, these must be deleted manually from the genomic sequences and restart the positive selection with the truncated sequences


